utils/ofi_tcp_keepalive.patch

```diff
@@ -0,0 +1,139 @@
+commit 7e590cfd30667b88e7755bfe265906ec19397d78
+Author: Di Wang <ddiwang@google.com>
+Date:   Thu May 8 18:41:22 2025 +0000
+
+    enable keepalive for libfabric
+    
+    Signed-off-by: Di Wang <ddiwang@google.com>
+
+diff --git a/prov/tcp/src/xnet.h b/prov/tcp/src/xnet.h
+index 6d2cbe029..eeeae5599 100644
+--- a/prov/tcp/src/xnet.h
++++ b/prov/tcp/src/xnet.h
+@@ -310,6 +310,9 @@ ssize_t xnet_get_conn(struct xnet_rdm *rdm, fi_addr_t dest_addr,
+ struct xnet_ep *xnet_get_rx_ep(struct xnet_rdm *rdm, fi_addr_t addr);
+ void xnet_freeall_conns(struct xnet_rdm *rdm);
+ 
++int xnet_enable_keepalive(struct xnet_ep *ep);
++int xnet_disable_keepalive(struct xnet_ep *ep);
++
+ struct xnet_uring {
+ 	struct fid fid;
+ 	ofi_io_uring_t ring;
+@@ -774,8 +777,8 @@ do {						\
+ 	ofi_straddr(_src_buf, &_src_len, ofi_translate_addr_format(ofi_sa_family(_src_addr)),	\
+ 		    _src_addr);									\
+ 												\
+-	FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "error on %s From %s : %s\n", _buf, _src_buf,	\
+-		fi_strerror(err));								\
++	FI_WARN(&xnet_prov, FI_LOG_EP_CTRL, "ep %p state %u error on %s From %s : %s\n",	\
++		ep, ep->state, _buf, _src_buf, fi_strerror(err));				\
+ } while(0)
+ 
+ #endif //_XNET_H_
+diff --git a/prov/tcp/src/xnet_cm.c b/prov/tcp/src/xnet_cm.c
+index 1ae35d27d..64d66922a 100644
+--- a/prov/tcp/src/xnet_cm.c
++++ b/prov/tcp/src/xnet_cm.c
+@@ -203,6 +203,8 @@ void xnet_req_done(struct xnet_ep *ep)
+ 	ret = xnet_req_done_internal(ep);
+ 	if (ret)
+ 		goto disable;
++
++	xnet_disable_keepalive(ep);
+ 	return;
+ 
+ disable:
+diff --git a/prov/tcp/src/xnet_ep.c b/prov/tcp/src/xnet_ep.c
+index 3fc566b7d..9f26d9c63 100644
+--- a/prov/tcp/src/xnet_ep.c
++++ b/prov/tcp/src/xnet_ep.c
+@@ -156,6 +156,66 @@ static void xnet_set_no_port(SOCKET sock)
+ #define xnet_set_no_port(sock)
+ #endif
+ 
++int
++xnet_disable_keepalive(struct xnet_ep *ep)
++{
++	int optval = 0;
++	int ret;
++
++	ret = setsockopt(ep->bsock.sock, SOL_SOCKET, SO_KEEPALIVE, (char *)&optval,
```

Comment:
nic: should be `(const void *)&optval`.

Reply:
usually you don't need cast if target is void *

Reply:
ah, sure